### PR TITLE
TensorflowTrainer improvements

### PR DESCRIPTION
**Pull request by andrewyates**
from @crystina-z's branch:
- change the meaning of itersize in tensorflow trainer from "number of steps" to "number of instance" (as mentioned in #117 )
- add fast-forward training to tensorflow trainer (review required)
- rename warmupsteps -> warmupiters, decaystep -> decayiters
- change the criteria of determinating "bert" variable, to also include "electra"

Additional changes:
- Updated `itersize` in docs/reproduction to keep behavior the same
- Reverted to previous linear decay formula to keep behavior the same. (Also: new one had a bug that could make LR negative.)
- Added `reranker.trainer.amp` config option defaulting to false (same as pytorch)
- Fix TF loss averaging in logging statement, which was also broken previously
- Add previous tfrecord fixes (@crystina-z can you check these to make sure I merged correctly?)

### CODE REVIEW

File: capreolus/trainer/tensorflow.py [hunk 1]
```diff
@@ -427,34 +427,33 @@ def convert_to_tf_dev_record(self, reranker, dataset):
...
 
         # TPU's require drop_remainder = True. But we cannot drop things from validation dataset
         # As a workaroud, we pad the dataset with the last sample until it reaches the batch size.
         if len(tf_features) % evalbatch:
             num_elements_to_add = evalbatch - (len(tf_features) % evalbatch)
             logger.debug("Number of elements to add in the last batch: {}".format(num_elements_to_add))
             element_to_copy = tf_features[-1]
-            for i in range(num_elements_to_add):
+            for i in range(self.config["evalbatch"]):
                 tf_features.append(copy(element_to_copy))
```
Comment (andrewyates) [inline capreolus/trainer/tensorflow.py:447]: @crystina-z is this correct?

File: capreolus/trainer/tensorflow.py [hunk 2]
```diff
@@ -81,11 +104,16 @@ def validate(self):
 
     def train(self, reranker, train_dataset, train_output_path, dev_data, dev_output_path, qrels, metric, relevance_level=1):
         if self.tpu:
-            train_output_path = "{0}/{1}/{2}".format(
-                self.config["storage"], "train_output", hashlib.md5(str(train_output_path).encode("utf-8")).hexdigest()
+            # WARNING: not sure if pathlib is compatible with gs://
```
Comment (crystina-z) [inline capreolus/trainer/tensorflow.py:107]: just convert it to `Path` should be fine (if we not trying to do `p.mkdir` etc)
```
>> Path("gs://bucket") / "sample_dir"
PosixPath('gs:/bucket/sample_dir')
```

File: capreolus/trainer/tensorflow.py (hunk 1 above)
Comment (crystina-z) [reply capreolus/trainer/tensorflow.py:449]: I just realized maybe here we should always append regardless of how many samples we have in `tf_features` (unless we keep tracking the *total* number of dev samples we wrote to tfrecord).
